### DISCUSSION TITLE
Make Sphere Shape Differentiable?

### DISCUSSION THREAD

**Discussion by kimren227** [mitsuba-renderer/mitsuba3#231] [Feature requests]
## Summary

Thanks for open sourcing this amazing project! 

I wish to optimize the parameters of the sphere shape(sphere center, radius, etc) through inverse rendering. However, after modifying the "object pose estimation" tutorial to use sphere shapes instead of ply, I got "TypeError, expected a differentiable array". A quick go-through of the "sphere.cpp" seems to confirm that (no differentiable flag on sphere radius and center).

Is it possible to make them differentiable and discontinuous? If so, what's the recommended way of implementing it?

Thanks very much.

Daxuan

**Comment by Speierers** [comment]
Hi @kimren227,

Indeed this is a limitation currently in Mitsuba 3. It is on our TODO list and we will probably take a look at this in the near future.

If you want to give this a try, I would start by changing the flag using in `Sphere::traverse()` to make the `to_world` parameter differentiable (and discontinuous). Then you will need to re-write the `Sphere::update()` method to make sure the the updated members (e.g. `m_radius`, `m_center`, ...) remain differentiable w.r.t. `to_world`. This could be achieved using `dr::replace_grad()` for instance.

**Comment by Speierers** [comment]
I will close this discussion as it is already discussed [here](https://github.com/mitsuba-renderer/mitsuba3/discussions/151).
